# **KPMG VIRTUAL INTERNSHIP PROJECT**

**TASK 1: DATA QUALITY ASSESSMENT**


The client provided KPMG with 4 datasets:

1.Customer Demographic

2.Customer Addresses

3.Transactions data in the past 3 months

4.New Customer List

In [ ]:
#Importing Libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns


In [ ]:
#LOADING THE DATASET
data = pd.ExcelFile('/content/KPMG_VI_New_raw_data_update_final.xlsx')

In [ ]:
#Reading each file separately
Transactions = pd.read_excel(data, 'Transactions')
NewCustomerList = pd.read_excel(data, 'NewCustomerList')
CustomerDemographic = pd.read_excel(data, 'CustomerDemographic')
CustomerAddress = pd.read_excel(data, 'CustomerAddress')

<ipython-input-3-ec2de8ff8fd5>:3: FutureWarning: Inferring datetime64[ns] from data containing strings is deprecated and will be removed in a future version. To retain the old behavior explicitly pass Series(data, dtype=datetime64[ns])
  NewCustomerList = pd.read_excel(data, 'NewCustomerList')
<ipython-input-3-ec2de8ff8fd5>:4: FutureWarning: Inferring datetime64[ns] from data containing strings is deprecated and will be removed in a future version. To retain the old behavior explicitly pass Series(data, dtype=datetime64[ns])
  CustomerDemographic = pd.read_excel(data, 'CustomerDemographic')


**EXPLORING THE 'TRANSACTIONS' DATASET**

In [ ]:
# Dataset First Look
Transactions.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  product_first_sold_date  
0       71.49          53.62                  41245.0  
1     2091.47         388.92                  41701.0  
2     1793.43         248.82                  36361.0  
3     1198.46         381.10                  36145.0  
4     1765.30         709.48                  42226.0

In [ ]:
# Dataset Rows & Columns count
Transactions.shape

(20000, 13)

There are 20000 rows and 13 columns in the transaction dataset.

In [ ]:
# Dataset Info
Transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           20000 non-null  int64         
 1   product_id               20000 non-null  int64         
 2   customer_id              20000 non-null  int64         
 3   transaction_date         20000 non-null  datetime64[ns]
 4   online_order             19640 non-null  float64       
 5   order_status             20000 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               20000 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

In [ ]:
# Dataset Duplicate Value Count
Transactions.duplicated().sum()

0

We can see that there are no duplicated values in the transaction dataset.

In [ ]:
# Missing Values/Null Values Count
Transactions.isnull().sum()

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               360
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

There are 360 null values in the column 'online_order' and 197 null values in the columns 'brand', 'product_line', 'product_class', 'product_size', 'standard_cost' and 'product_first_sold_date'.

---
we must replace or remove the null values as per the nature of the columns.


In [ ]:
# Dataset Describe
Transactions.describe(include='all')

<ipython-input-10-5d477a6ce7df>:2: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  Transactions.describe(include='all')


transaction_id   product_id   customer_id     transaction_date  \
count     20000.000000  20000.00000  20000.000000                20000   
unique             NaN          NaN           NaN                  364   
top                NaN          NaN           NaN  2017-02-14 00:00:00   
freq               NaN          NaN           NaN                   82   
first              NaN          NaN           NaN  2017-01-01 00:00:00   
last               NaN          NaN           NaN  2017-12-30 00:00:00   
mean      10000.500000     45.36465   1738.246050                  NaN   
std        5773.647028     30.75359   1011.951046                  NaN   
min           1.000000      0.00000      1.000000                  NaN   
25%        5000.750000     18.00000    857.750000                  NaN   
50%       10000.500000     44.00000   1736.000000                  NaN   
75%       15000.250000     72.00000   2613.000000                  NaN   
max       20000.000000    100.00000   5034.000000                  NaN   

        online_order order_status  brand product_line product_class  \
count   19640.000000        20000  19803        19803         19803   
unique           NaN            2      6            4             3   
top              NaN     Approved  Solex     Standard        medium   
freq             NaN        19821   4253        14176         13826   
first            NaN          NaN    NaN          NaN           NaN   
last             NaN          NaN    NaN          NaN           NaN   
mean        0.500458          NaN    NaN          NaN           NaN   
std         0.500013          NaN    NaN          NaN           NaN   
min         0.000000          NaN    NaN          NaN           NaN   
25%         0.000000          NaN    NaN          NaN           NaN   
50%         1.000000          NaN    NaN          NaN           NaN   
75%         1.000000          NaN    NaN          NaN           NaN   
max         1.000000          NaN    NaN          NaN           NaN   

       product_size    list_price  standard_cost  product_first_sold_date  
count         19803  20000.000000   19803.000000             19803.000000  
unique            3           NaN            NaN                      NaN  
top          medium           NaN            NaN                      NaN  
freq          12990           NaN            NaN                      NaN  
first           NaN           NaN            NaN                      NaN  
last            NaN           NaN            NaN                      NaN  
mean            NaN   1107.829449     556.046951             38199.776549  
std             NaN    582.825242     405.955660              2875.201110  
min             NaN     12.010000       7.210000             33259.000000  
25%             NaN    575.270000     215.140000             35667.000000  
50%             NaN   1163.890000     507.580000             38216.000000  
75%             NaN   1635.300000     795.100000             40672.000000  
max             NaN   2091.470000    1759.850000             42710.000000

In [ ]:
#Looking at number of columns in ths dataset
Transactions.columns

Index(['transaction_id', 'product_id', 'customer_id', 'transaction_date',
       'online_order', 'order_status', 'brand', 'product_line',
       'product_class', 'product_size', 'list_price', 'standard_cost',
       'product_first_sold_date'],
      dtype='object')

Let's Explore few of the columns in the Transaction Dataset

In [ ]:
T_col=['order_status','brand','product_line','product_class', 'product_size',]
for i in T_col:
  
  print(f'The unique values in {i} are:', Transactions[i].unique())

The unique values in order_status are: ['Approved' 'Cancelled']
The unique values in brand are: ['Solex' 'Trek Bicycles' 'OHM Cycles' 'Norco Bicycles' 'Giant Bicycles'
 'WeareA2B' nan]
The unique values in product_line are: ['Standard' 'Road' 'Mountain' 'Touring' nan]
The unique values in product_class are: ['medium' 'low' 'high' nan]
The unique values in product_size are: ['medium' 'large' 'small' nan]


The null values in these columns should be replace or remove as per the nature of the columns.

In [ ]:
Transactions['order_status'].value_counts()

Approved     19821
Cancelled      179
Name: order_status, dtype: int64

In [ ]:
Transactions['brand'].value_counts()

Solex             4253
Giant Bicycles    3312
WeareA2B          3295
OHM Cycles        3043
Trek Bicycles     2990
Norco Bicycles    2910
Name: brand, dtype: int64

In [ ]:
Transactions['product_line'].value_counts()

Standard    14176
Road         3970
Touring      1234
Mountain      423
Name: product_line, dtype: int64

In [ ]:
Transactions['product_class'].value_counts()

medium    13826
high       3013
low        2964
Name: product_class, dtype: int64

In [ ]:
Transactions['product_size'].value_counts()

medium    12990
large      3976
small      2837
Name: product_size, dtype: int64

In [ ]:
Transactions['product_first_sold_date']

0        41245.0
1        41701.0
2        36361.0
3        36145.0
4        42226.0
          ...   
19995    37823.0
19996    35560.0
19997    40410.0
19998    38216.0
19999    36334.0
Name: product_first_sold_date, Length: 20000, dtype: float64

In [ ]:
#convert date column from integer to datetime
Transactions['product_first_sold_date'] = pd.to_datetime(Transactions['product_first_sold_date'], unit='s')
Transactions['product_first_sold_date'].head()

0   1970-01-01 11:27:25
1   1970-01-01 11:35:01
2   1970-01-01 10:06:01
3   1970-01-01 10:02:25
4   1970-01-01 11:43:46
Name: product_first_sold_date, dtype: datetime64[ns]

In [ ]:
Transactions['product_first_sold_date'].head(10)

0   1970-01-01 11:27:25
1   1970-01-01 11:35:01
2   1970-01-01 10:06:01
3   1970-01-01 10:02:25
4   1970-01-01 11:43:46
5   1970-01-01 10:50:31
6   1970-01-01 09:29:25
7   1970-01-01 11:05:15
8   1970-01-01 09:17:35
9   1970-01-01 10:36:56
Name: product_first_sold_date, dtype: datetime64[ns]

The values in the product_first_sold_date columns are not correct as it shows everything happening the same day at different times.

**EXPLORING THE 'NewCustomerList' DATASET**

In [ ]:
# Dataset First Look
NewCustomerList.head()

first_name  last_name  gender  past_3_years_bike_related_purchases  \
0    Chickie    Brister    Male                                   86   
1      Morly     Genery    Male                                   69   
2    Ardelis  Forrester  Female                                   10   
3     Lucine      Stutt  Female                                   64   
4    Melinda     Hadlee  Female                                   34   

         DOB                   job_title job_industry_category  \
0 1957-07-12             General Manager         Manufacturing   
1 1970-03-22         Structural Engineer              Property   
2 1974-08-28      Senior Cost Accountant    Financial Services   
3 1979-01-28  Account Representative III         Manufacturing   
4 1965-09-21           Financial Analyst    Financial Services   

      wealth_segment deceased_indicator owns_car  ...  state    country  \
0      Mass Customer                  N      Yes  ...    QLD  Australia   
1      Mass Customer                  N       No  ...    NSW  Australia   
2  Affluent Customer                  N       No  ...    VIC  Australia   
3  Affluent Customer                  N      Yes  ...    QLD  Australia   
4  Affluent Customer                  N       No  ...    NSW  Australia   

   property_valuation Unnamed: 16 Unnamed: 17  Unnamed: 18  Unnamed: 19  \
0                   6        0.76      0.9500       1.1875     1.009375   
1                  11        0.48      0.4800       0.6000     0.510000   
2                   5        0.85      0.8500       0.8500     0.850000   
3                   1        0.43      0.5375       0.5375     0.537500   
4                   9        0.77      0.7700       0.9625     0.962500   

   Unnamed: 20  Rank     Value  
0            1     1  1.718750  
1            1     1  1.718750  
2            1     1  1.718750  
3            4     4  1.703125  
4            4     4  1.703125  

[5 rows x 23 columns]

In [ ]:
# Dataset Rows & Columns count
NewCustomerList.shape

(1000, 23)

In [ ]:
# Made a copy of original dataset for doing feature engineering
NewCustomerList_c = NewCustomerList.copy()

In [ ]:
#Dropping the unnamed columns
NewCustomerList_c.drop(['Unnamed: 16', 'Unnamed: 17', 'Unnamed: 18',
       'Unnamed: 19', 'Unnamed: 20'], axis=1, inplace=True)


In [ ]:
# Dataset Info
NewCustomerList_c.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 18 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   first_name                           1000 non-null   object        
 1   last_name                            971 non-null    object        
 2   gender                               1000 non-null   object        
 3   past_3_years_bike_related_purchases  1000 non-null   int64         
 4   DOB                                  983 non-null    datetime64[ns]
 5   job_title                            894 non-null    object        
 6   job_industry_category                835 non-null    object        
 7   wealth_segment                       1000 non-null   object        
 8   deceased_indicator                   1000 non-null   object        
 9   owns_car                             1000 non-null   object        
 10  tenure       

In [ ]:
# Dataset Duplicate Value Count
NewCustomerList_c.duplicated().sum()

0

In [ ]:
# Missing Values/Null Values Count
NewCustomerList_c.isnull().sum()

first_name                               0
last_name                               29
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     17
job_title                              106
job_industry_category                  165
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                   0
address                                  0
postcode                                 0
state                                    0
country                                  0
property_valuation                       0
Rank                                     0
Value                                    0
dtype: int64

In [ ]:
# Dataset Describe
NewCustomerList_c.describe(include='all')

<ipython-input-48-d09ab2acafbe>:2: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  NewCustomerList_c.describe(include='all')


first_name last_name  gender  past_3_years_bike_related_purchases  \
count        1000       971    1000                          1000.000000   
unique        940       961       3                                  NaN   
top      Rozamond    Sissel  Female                                  NaN   
freq            3         2     513                                  NaN   
first         NaN       NaN     NaN                                  NaN   
last          NaN       NaN     NaN                                  NaN   
mean          NaN       NaN     NaN                            49.836000   
std           NaN       NaN     NaN                            27.796686   
min           NaN       NaN     NaN                             0.000000   
25%           NaN       NaN     NaN                            26.750000   
50%           NaN       NaN     NaN                            51.000000   
75%           NaN       NaN     NaN                            72.000000   
max           NaN       NaN     NaN                            99.000000   

                        DOB            job_title job_industry_category  \
count                   983                  894                   835   
unique                  958                  184                     9   
top     1998-02-05 00:00:00  Associate Professor    Financial Services   
freq                      2                   15                   203   
first   1938-06-08 00:00:00                  NaN                   NaN   
last    2002-02-27 00:00:00                  NaN                   NaN   
mean                    NaN                  NaN                   NaN   
std                     NaN                  NaN                   NaN   
min                     NaN                  NaN                   NaN   
25%                     NaN                  NaN                   NaN   
50%                     NaN                  NaN                   NaN   
75%                     NaN                  NaN                   NaN   
max                     NaN                  NaN                   NaN   

       wealth_segment deceased_indicator owns_car       tenure  \
count            1000               1000     1000  1000.000000   
unique              3                  1        2          NaN   
top     Mass Customer                  N       No          NaN   
freq              508               1000      507          NaN   
first             NaN                NaN      NaN          NaN   
last              NaN                NaN      NaN          NaN   
mean              NaN                NaN      NaN    11.388000   
std               NaN                NaN      NaN     5.037145   
min               NaN                NaN      NaN     0.000000   
25%               NaN                NaN      NaN     7.000000   
50%               NaN                NaN      NaN    11.000000   
75%               NaN                NaN      NaN    15.000000   
max               NaN                NaN      NaN    22.000000   

                 address     postcode state    country  property_valuation  \
count               1000  1000.000000  1000       1000         1000.000000   
unique              1000          NaN     3          1                 NaN   
top     45 Shopko Center          NaN   NSW  Australia                 NaN   
freq                   1          NaN   506       1000                 NaN   
first                NaN          NaN   NaN        NaN                 NaN   
last                 NaN          NaN   NaN        NaN                 NaN   
mean                 NaN  3019.227000   NaN        NaN            7.397000   
std                  NaN   848.895767   NaN        NaN            2.758804   
min                  NaN  2000.000000   NaN        NaN            1.000000   
25%                  NaN  2209.000000   NaN        NaN            6.000000   
50%                  NaN  2800.000000   NaN        NaN            8.000000   
75%                  NaN  3845.500000   NaN 

In [ ]:
#Looking at different columns in the dataset
NewCustomerList_c.columns

Index(['first_name', 'last_name', 'gender',
       'past_3_years_bike_related_purchases', 'DOB', 'job_title',
       'job_industry_category', 'wealth_segment', 'deceased_indicator',
       'owns_car', 'tenure', 'address', 'postcode', 'state', 'country',
       'property_valuation', 'Rank', 'Value'],
      dtype='object')

Let's explore the different columns in the 'New Customer List' dataset.

In [ ]:
NewCustomerList_c['gender'].value_counts()

Female    513
Male      470
U          17
Name: gender, dtype: int64

In [ ]:
#Counting Unspecified/unknown genders
count=0
for i in NewCustomerList.gender:
  if i == "U":
    count+=1
  else:
    pass
print(count)

17


There are 17 rows in the gender column with unknown/unspecified genders.

In [ ]:
NewCustomerList_c['DOB'].value_counts()

1998-02-05    2
1978-01-15    2
1977-11-08    2
1951-11-28    2
1979-07-28    2
             ..
1945-08-08    1
1943-08-27    1
1999-10-24    1
1976-01-24    1
1955-10-02    1
Name: DOB, Length: 958, dtype: int64

We can see that few DOB values are not correct which should be replaced or removed

In [ ]:
NewCustomerList['job_industry_category'].value_counts()

Financial Services    203
Manufacturing         199
Health                152
Retail                 78
Property               64
IT                     51
Entertainment          37
Argiculture            26
Telecommunications     25
Name: job_industry_category, dtype: int64

In [ ]:
NewCustomerList['wealth_segment'].value_counts()

Mass Customer        508
High Net Worth       251
Affluent Customer    241
Name: wealth_segment, dtype: int64

In [ ]:
NewCustomerList['state'].value_counts()

NSW    506
VIC    266
QLD    228
Name: state, dtype: int64

In [ ]:
NewCustomerList['owns_car'].value_counts()

No     507
Yes    493
Name: owns_car, dtype: int64

In [ ]:
NewCustomerList['deceased_indicator'].value_counts()

N    1000
Name: deceased_indicator, dtype: int64

**EXPLORING THE 'CustomerDemographic' DATASET**

In [ ]:
# Dataset First Look
CustomerDemographic.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  \
0                Health      Mass Customer                  N   
1    Financial Services      Mass Customer                  N   
2              Property      Mass Customer                  N   
3                    IT      Mass Customer                  N   
4                   NaN  Affluent Customer                  N   

                                             default owns_car  tenure  
0                                                 "'      Yes    11.0  
1                       <script>alert('hi')</script>      Yes    16.0  
2                                2018-02-01 00:00:00      Yes    15.0  
3  () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                NIL      Yes     8.0

In [ ]:
# Dataset Rows & Columns count
CustomerDemographic.shape

(4000, 13)

There are 40000 rows and 13 columns in the customer demographic dataset.

In [ ]:
# Dataset Info
CustomerDemographic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          4000 non-null   int64         
 1   first_name                           4000 non-null   object        
 2   last_name                            3875 non-null   object        
 3   gender                               4000 non-null   object        
 4   past_3_years_bike_related_purchases  4000 non-null   int64         
 5   DOB                                  3913 non-null   datetime64[ns]
 6   job_title                            3494 non-null   object        
 7   job_industry_category                3344 non-null   object        
 8   wealth_segment                       4000 non-null   object        
 9   deceased_indicator                   4000 non-null   object        
 10  default     

In [ ]:
# Dataset Duplicate Value Count
CustomerDemographic.duplicated().sum()

0

No duplicate values in this dataset.

In [ ]:
# Missing Values/Null Values Count
CustomerDemographic.isnull().sum()

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
default                                302
owns_car                                 0
tenure                                  87
dtype: int64

There are 125, 87, 506, 656, 302, 87 null values in the last_name, DOB, job-title, job_industry_category, default, tenure respectively.

In [ ]:
# Dataset Describe
CustomerDemographic.describe(include='all')

<ipython-input-60-97b1eabe7a8c>:2: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  CustomerDemographic.describe(include='all')


customer_id first_name last_name  gender  \
count   4000.000000       4000      3875    4000   
unique          NaN       3139      3725       6   
top             NaN        Max  Pristnor  Female   
freq            NaN          5         3    2037   
first           NaN        NaN       NaN     NaN   
last            NaN        NaN       NaN     NaN   
mean    2000.500000        NaN       NaN     NaN   
std     1154.844867        NaN       NaN     NaN   
min        1.000000        NaN       NaN     NaN   
25%     1000.750000        NaN       NaN     NaN   
50%     2000.500000        NaN       NaN     NaN   
75%     3000.250000        NaN       NaN     NaN   
max     4000.000000        NaN       NaN     NaN   

        past_3_years_bike_related_purchases                  DOB  \
count                           4000.000000                 3913   
unique                                  NaN                 3448   
top                                     NaN  1978-01-30 00:00:00   
freq                                    NaN                    7   
first                                   NaN  1843-12-21 00:00:00   
last                                    NaN  2002-03-11 00:00:00   
mean                              48.890000                  NaN   
std                               28.715005                  NaN   
min                                0.000000                  NaN   
25%                               24.000000                  NaN   
50%                               48.000000                  NaN   
75%                               73.000000                  NaN   
max                               99.000000                  NaN   

                                   job_title job_industry_category  \
count                                   3494                  3344   
unique                                   195                     9   
top     Business Systems Development Analyst         Manufacturing   
freq                                      45                   799   
first                                    NaN                   NaN   
last                                     NaN                   NaN   
mean                                     NaN                   NaN   
std                                      NaN                   NaN   
min                                      NaN                   NaN   
25%                                      NaN                   NaN   
50%                                      NaN                   NaN   
75%                                      NaN                   NaN   
max                                      NaN                   NaN   

       wealth_segment deceased_indicator  default owns_car       tenure  
count            4000               4000   3698.0     4000  3913.000000  
unique              3                  2     90.0        2          NaN  
top     Mass Customer                  N    100.0      Yes          NaN  
freq             2000               3998    113.0     2024          NaN  
first             NaN                NaN      NaN      NaN          NaN  
last              NaN                NaN      NaN      NaN          NaN  
mean              NaN                NaN      NaN      NaN    10.657041  
std               NaN                NaN      NaN      NaN     5.660146  
min               NaN                NaN      NaN      NaN     1.000000  
25%               NaN                NaN      NaN      NaN     6.000000  
50%               NaN                NaN      NaN      NaN    11.000000  
75%               NaN                NaN      NaN      NaN    15.000000  
max               NaN                NaN      NaN      NaN    22.000000

In [ ]:
#Exploring few columns to avoid any possible errors
CD_col=['gender','job_industry_category','wealth_segment','deceased_indicator', 'owns_car']
for i in CD_col:
  
  print(f'The unique values in {i} are:', CustomerDemographic[i].unique())

The unique values in gender are: ['F' 'Male' 'Female' 'U' 'Femal' 'M']
The unique values in job_industry_category are: ['Health' 'Financial Services' 'Property' 'IT' nan 'Retail' 'Argiculture'
 'Manufacturing' 'Telecommunications' 'Entertainment']
The unique values in wealth_segment are: ['Mass Customer' 'Affluent Customer' 'High Net Worth']
The unique values in deceased_indicator are: ['N' 'Y']
The unique values in owns_car are: ['Yes' 'No']


Possible issues as mentinoed below:


1.   Too many words were in use to represent 'Male' and 'Female' gender which must be changed.
2.   'job_industry_category' this column contains null values which should be handle as per the nature of the column.



In [ ]:
# Made a copy of original dataset for doing feature engineering
CustomerDemographic_c = CustomerDemographic.copy()

Let's explore different coulmns in the 'Customer Demographic' dataset

In [ ]:
#Looking at different columns in the dataset
CustomerDemographic_c.columns

Index(['customer_id', 'first_name', 'last_name', 'gender',
       'past_3_years_bike_related_purchases', 'DOB', 'job_title',
       'job_industry_category', 'wealth_segment', 'deceased_indicator',
       'default', 'owns_car', 'tenure', 'age'],
      dtype='object')

In [ ]:
CustomerDemographic_c['gender'].value_counts()

Female    2037
Male      1872
U           88
F            1
Femal        1
M            1
Name: gender, dtype: int64

Certain categories are not correctly titled. The names in these categories should be re-named.

In [ ]:
#Re-naming the categories
CustomerDemographic_c['gender'] = CustomerDemographic_c['gender'].replace('F','Female').replace('M','Male').replace('Femal','Female').replace('U','Unknown')

In [ ]:
CustomerDemographic_c['gender'].value_counts()

Female     2039
Male       1873
Unknown      88
Name: gender, dtype: int64

In [ ]:
CustomerDemographic_c['DOB'].value_counts()

1978-01-30    7
1964-07-08    4
1962-12-17    4
1978-08-19    4
1977-05-13    4
             ..
1989-06-16    1
1998-09-30    1
1985-03-11    1
1989-10-23    1
1991-11-05    1
Name: DOB, Length: 3448, dtype: int64

In [ ]:
# sorting date with sort_value() function
Final_result = CustomerDemographic_c.sort_values('DOB')
  
print(" Sorted Date DataFrame")
Final_result

 Sorted Date DataFrame


customer_id first_name  last_name   gender  \
33             34   Jephthah   Bachmann  Unknown   
719           720     Darrel      Canet     Male   
1091         1092     Katlin    Creddon   Female   
3409         3410    Merrili    Brittin   Female   
2412         2413      Abbey     Murrow     Male   
...           ...        ...        ...      ...   
3778         3779      Ulick     Daspar  Unknown   
3882         3883      Nissa     Conrad  Unknown   
3930         3931      Kylie      Epine  Unknown   
3934         3935     Teodor  Alfonsini  Unknown   
3997         3998     Sarene    Woolley  Unknown   

      past_3_years_bike_related_purchases        DOB  \
33                                     59 1843-12-21   
719                                    67 1931-10-23   
1091                                   56 1935-08-22   
3409                                   93 1940-09-22   
2412                                   27 1943-08-11   
...                                   ...        ...   
3778                                   68        NaT   
3882                                   35        NaT   
3930                                   19        NaT   
3934                                   72        NaT   
3997                                   60        NaT   

                     job_title job_industry_category     wealth_segment  \
33             Legal Assistant                    IT  Affluent Customer   
719         Recruiting Manager                Retail  Affluent Customer   
1091        VP Quality Control                Retail      Mass Customer   
3409                       NaN              Property      Mass Customer   
2412  Environmental Specialist         Manufacturing     High Net Worth   
...                        ...                   ...                ...   
3778                       NaN                    IT  Affluent Customer   
3882           Legal Assistant                    IT      Mass Customer   
3930                       NaN                    IT     High Net Worth   
3934                       NaN                    IT     High Net Worth   
3997         Assistant Manager                    IT     High Net Worth   

     deceased_indicator             default owns_car  tenure    age  
33                    N                 NaN       No    20.0  180.0  
719                   N  1;DROP TABLE users       No     6.0   92.0  
1091                  N  ì¬íê³¼íì ì´íì°êµ¬ì       No     5.0   88.0  
3409                  N                   á       No    16.0   83.0  
2412                  N                  á       Yes    17.0   80.0  
...                 ...                 ...      ...     ...    ...  
3778                  N                 NaN       No     NaN    NaN  
3882                  N                 NaN       No     NaN    NaN  
3930                  N                 NaN      Yes     NaN    NaN  
3934                  N                 NaN      Yes     NaN    NaN  
3997                  N                 NaN       No     NaN    NaN  

[4000 rows x 14 columns]

We can see that the DOB of first customer is 1843-12-21 which is quite starnge. So, we need to take appropriate action on this case.

In [ ]:
CustomerDemographic_c['job_title'].value_counts()

Business Systems Development Analyst    45
Tax Accountant                          44
Social Worker                           44
Internal Auditor                        42
Recruiting Manager                      41
                                        ..
Database Administrator I                 4
Health Coach I                           3
Health Coach III                         3
Research Assistant III                   3
Developer I                              1
Name: job_title, Length: 195, dtype: int64

In [ ]:
CustomerDemographic_c['job_industry_category'].value_counts()

Manufacturing         799
Financial Services    774
Health                602
Retail                358
Property              267
IT                    223
Entertainment         136
Argiculture           113
Telecommunications     72
Name: job_industry_category, dtype: int64

In [ ]:
CustomerDemographic_c['wealth_segment'].value_counts()

Mass Customer        2000
High Net Worth       1021
Affluent Customer     979
Name: wealth_segment, dtype: int64

In [ ]:
CustomerDemographic_c['deceased_indicator'].value_counts()

N    3998
Y       2
Name: deceased_indicator, dtype: int64

In [ ]:
CustomerDemographic_c['default'].value_counts()

100                                       113
1                                         112
-1                                        111
-100                                       99
١٢٣                                     53
                                         ... 
testâ testâ«                               31
/dev/null; touch /tmp/blns.fail ; echo     30
âªâªtest⪠                                29
ì¸ëë°í ë¥´                                 27
,ãã»:*:ã»ãâ( â» Ï â» )ãã»:*:ã»ãâ           25
Name: default, Length: 90, dtype: int64

This column is completely useless which must be dropped from the dataset in order to minimize the confusion.

In [ ]:
CustomerDemographic_c.drop('default', axis=1)

customer_id      first_name  last_name   gender  \
0               1         Laraine  Medendorp   Female   
1               2             Eli    Bockman     Male   
2               3           Arlin     Dearle     Male   
3               4          Talbot        NaN     Male   
4               5  Sheila-kathryn     Calton   Female   
...           ...             ...        ...      ...   
3995         3996         Rosalia   Halgarth   Female   
3996         3997          Blanch     Nisuis   Female   
3997         3998          Sarene    Woolley  Unknown   
3998         3999       Patrizius        NaN     Male   
3999         4000           Kippy    Oldland     Male   

      past_3_years_bike_related_purchases        DOB               job_title  \
0                                      93 1953-10-12     Executive Secretary   
1                                      81 1980-12-16  Administrative Officer   
2                                      61 1954-01-20      Recruiting Manager   
3                                      33 1961-10-03                     NaN   
4                                      56 1977-05-13           Senior Editor   
...                                   ...        ...                     ...   
3995                                    8 1975-08-09   VP Product Management   
3996                                   87 2001-07-13         Statistician II   
3997                                   60        NaT       Assistant Manager   
3998                                   11 1973-10-24                     NaN   
3999                                   76 1991-11-05    Software Engineer IV   

     job_industry_category     wealth_segment deceased_indicator owns_car  \
0                   Health      Mass Customer                  N      Yes   
1       Financial Services      Mass Customer                  N      Yes   
2                 Property      Mass Customer                  N      Yes   
3                       IT      Mass Customer                  N       No   
4                      NaN  Affluent Customer                  N      Yes   
...                    ...                ...                ...      ...   
3995                Health      Mass Customer                  N       No   
3996         Manufacturing     High Net Worth                  N      Yes   
3997                    IT     High Net Worth                  N       No   
3998         Manufacturing  Affluent Customer                  N      Yes   
3999                   NaN  Affluent Customer                  N       No   

      tenure   age  
0       11.0  70.0  
1       16.0  43.0  
2       15.0  69.0  
3        7.0  62.0  
4        8.0  46.0  
...      ...   ...  
3995    19.0  48.0  
3996     1.0  22.0  
3997     NaN   NaN  
3998    10.0  50.0  
3999    11.0  32.0  

[4000 rows x 13 columns]

In [ ]:
CustomerDemographic_c['owns_car'].value_counts()

Yes    2024
No     1976
Name: owns_car, dtype: int64

**EXPLORING THE 'CustomerAddress' DATASET**

In [ ]:
# Dataset First Look
CustomerAddress.head()

customer_id              address  postcode            state    country  \
0            1   060 Morning Avenue      2016  New South Wales  Australia   
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211              QLD  Australia   
3            5  17979 Del Mar Point      2448  New South Wales  Australia   
4            6     9 Oakridge Court      3216              VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

In [ ]:
# Dataset Rows & Columns count
CustomerAddress.shape

(3999, 6)

In [ ]:
# Dataset Info
CustomerAddress.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   int64 
 1   address             3999 non-null   object
 2   postcode            3999 non-null   int64 
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


In [ ]:
# Dataset Duplicate Value Count
CustomerAddress.duplicated().sum()

0

In [ ]:
# Missing Values/Null Values Count
CustomerAddress.isnull().sum()

customer_id           0
address               0
postcode              0
state                 0
country               0
property_valuation    0
dtype: int64

In [ ]:
# Dataset Describe
CustomerAddress.describe(include='all')

customer_id                  address     postcode state    country  \
count   3999.000000                     3999  3999.000000  3999       3999   
unique          NaN                     3996          NaN     5          1   
top             NaN  3 Mariners Cove Terrace          NaN   NSW  Australia   
freq            NaN                        2          NaN  2054       3999   
mean    2003.987997                      NaN  2985.755939   NaN        NaN   
std     1154.576912                      NaN   844.878364   NaN        NaN   
min        1.000000                      NaN  2000.000000   NaN        NaN   
25%     1004.500000                      NaN  2200.000000   NaN        NaN   
50%     2004.000000                      NaN  2768.000000   NaN        NaN   
75%     3003.500000                      NaN  3750.000000   NaN        NaN   
max     4003.000000                      NaN  4883.000000   NaN        NaN   

        property_valuation  
count          3999.000000  
unique                 NaN  
top                    NaN  
freq                   NaN  
mean              7.514379  
std               2.824663  
min               1.000000  
25%               6.000000  
50%               8.000000  
75%              10.000000  
max              12.000000

Let's explore the columns in 'Customer Address' dataset

In [ ]:
#Looking at different columns in the dataset
CustomerAddress.columns

Index(['customer_id', 'address', 'postcode', 'state', 'country',
       'property_valuation'],
      dtype='object')

In [ ]:
CustomerAddress['postcode'].value_counts()

2170    31
2155    30
2145    30
2153    29
3977    26
        ..
3808     1
3114     1
4721     1
4799     1
3089     1
Name: postcode, Length: 873, dtype: int64

In [ ]:
CustomerAddress['state'].value_counts()

NSW                2054
VIC                 939
QLD                 838
New South Wales      86
Victoria             82
Name: state, dtype: int64

In [ ]:
CustomerAddress['country'].value_counts()

Australia    3999
Name: country, dtype: int64

In [ ]:
CustomerAddress['property_valuation'].value_counts()

9     647
8     646
10    577
7     493
11    281
6     238
5     225
4     214
12    195
3     186
1     154
2     143
Name: property_valuation, dtype: int64

All the columns appear to have consistent and correct information.

### **THANK YOU**